principles.rst

```diff
@@ -168,26 +234,38 @@ Lexical Scoping Principle (LSP)
 **Principle**:
 
 a. For every appearance of an identifier,
-it is possible to determine whether that appearance is a *binding site* or an *occurrence* without examining the context.
+   it is possible to determine whether that appearance is a mere *occurrence*,
+   and thus must be bound elsewhere for the program to be valid,
+   or the variable is a *binding site* (or causes an implicit binding, which is close enough),
+   without examining the context.
 
 b. For every *occurrence* of an identifier,
-it is possible to uniquely identify its *binding site*, without involving the type system.
+   it is possible to uniquely identify its *binding site*, without involving the type system.
+
+This builds upon the `Explicit Binding Principle`_:
+whereas that former principle ensures that explicit alternatives to implicit binding constructs *exist at all*,
+this latter principle makes those explicit alternatives *compulsory*, because we must not have implicit binding in order to uphold this principle.
 
 The `Lexical Scoping Principle`_ is almost true today, with the following nuances:
 
-1. Template Haskell splices may need to be run before completing name resolution (and running those splices requires type-checking them).
+#. Template Haskell splices may need to be run before completing name resolution (and running those splices requires type-checking them).
 
-2. The `deprecated mechanism <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/exts/duplicate_record_fields.html#selector-functions>`_ for disambiguating duplicate record fields violates the `Lexical Scoping Principle`_ by requiring the type system.
+#. The `deprecated mechanism <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/exts/duplicate_record_fields.html#selector-functions>`_ for disambiguating duplicate record fields violates the `Lexical Scoping Principle`_ by requiring the type system.
 
-3. In a pattern signature,
+#. In a pattern signature,
    if we have ``f (x :: Maybe a)``,
    the ``a`` is an occurrence if ``a`` is already in scope,
-   and it is a binding site otherwise.
+   and is implicitly bound otherwise.
```

Comment:
See both replies below.

I don't mind putting it back the way it was, but I do think it's important for us deliberating the proposal to understand that even if isn't an "implicit binding", it *is* a violation of the **Explicit Binding Principle** because it is not explicit *and unambiguous*. (By which I mean the context doesn't matter as to whether it's a binding or not.)

Would it be better to rename this the **Explicit and Unambiguous Binding Principle**, or the **Context-free Binding Principle**?

Reply:
I must register here that your proposal makes the bindings more confusing to me than they currently are. You are free to stick to your guns there, it will however colour my recommendation.

To me the context-freeness was the purview of the lexical binding principle; while the presence of a binding site in the text was the domain of the explicit binding principle. You're making the distinction less clear to me, I can't quite understand your point of view yet. I may very well be missing something.

Reply:
Let me just recap your position as I understand it:

- Substantive changes seem mostly fine
   - substantive changes include `-XImplicitBinds` enabling pattern signature binds.
- Characterizing pattern signature binds as implicit binds seems bad or at least unnecessarily opinionated.
- Principles changes seem gratuitous
  - Principles changes are trying to demonstrate how pattern signature binds are dubious in the same way as the others.

So if you really want, I will relent and revert the principles changes in the interest of getting something done, but based on the above I hope it is clear why the amended proposal with the original principles feels shaky to me, and I worry how the committee would react.

If pattern signatures binds really are a perfectly good explicit binding, then it doesn't make sense that `-XImplicitBinds` would be required to use them! On the other hand if they are a dubvious implicit/contextual binding, then it does make sense that `-XImplictlBinds`/`-XContextualBinds` is a requirement to use them.

Ultimately, I don't really care what the names are, consider my proposal here subject to alpha equivalence. (That is why I bring up "contextual" vs "implicit".) I just want the principles and design decisions to live in harmony.

Also, I suspect you view might view the principles as a sort of append-only log, whereas I view the principles as a "living document" in which shuffling them around is a perfectly normal thing to do whether one is amending an existing proposal or writing a new proposal from scratch.

Reply:
> Substantive changes seem mostly fine
> 
>     * substantive changes include `-XImplicitBinds` enabling pattern signature binds.

My personal opinion on whether I agree with the change is insubstantial, honestly. This is why we have a committee. My first job as a shepherd is make sure that the proposal is clear and consistent. It is.

> Characterizing pattern signature binds as implicit binds seems bad or at least unnecessarily opinionated.

I'd even say that it's factually incorrect. It doesn't mean that they're not a bad idea though.

> Principles changes seem gratuitous
> 
>     * Principles changes are trying to demonstrate how pattern signature binds are dubious in the same way as the others.

My point is that the principles already did this. But you can stick to your guns, let's see what the rest of the committee says.

I'll be submitting for committee discussion now: I really should have done this last week as I'll be on holiday for the 3 next weeks and won't be checking my emails. The committee's decision will most likely be in January. I'm really sorry about this.

Reply:
OK sounds good. Basically I don't want to be so stubborn that the principles changes sink the rest of the proposal.

I take it you are saying the Lexical Scoping Principle covers this case today (I agree actually), so then should the extension be called `-XNonLexicalBinds` or `-XLexicalBinds`? I'll go with the flow on whatever the committee wants re naming.

Enjoy your vacation!